daffodil-core/src/main/scala/org/apache/daffodil/dsom/GroupRef.scala

```diff
@@ -53,24 +53,43 @@ trait GroupRef { self: ModelGroup =>
  * A GroupRefFactory (group reference) is an indirection to
  * factories to create a SequenceGroupRef, or ChoiceGroupRef.
  *
- * The refXMLArg is the xml for the group reference.
+ * The refXML is the xml for the group reference.
  *
  * This factory exists in order to make error messages refer to the
- * right part of the schema.
+ * right part of the schema and to navigate the ref prior to creating
+ * the object so that we can create a more specific kind of group ref that
+ * knows if it is referring to a choice or a sequence.
  */
-final class GroupRefFactory(refXMLArg: Node, val refLexicalParent: SchemaComponent, position: Int, isHidden: Boolean)
-  extends SchemaComponentFactory(refXMLArg, Some(refLexicalParent.schemaDocument))
+object GroupRefFactory {
+
+  def apply(refXML: Node, refLexicalParent: SchemaComponent, position: Int, isHidden: Boolean) = {
+    val f = new GroupRefFactory(refXML, refLexicalParent, position, isHidden)
+    f.groupRef
+  }
+}
+
+/**
+ * This is a quasi schema component. It is such simply to share a bunch of
+ * mechanism that schema components inherit, in a hassle free manner.
+ * (E.g., diagnostics mechanism that keeps track of file and line number).
+ */
+final class GroupRefFactory private (refXML: Node, val refLexicalParent: SchemaComponent, position: Int, isHidden: Boolean)
```

Comment:
For a class I think you have to put the private before the "(" of the arglist because this makes the constructor private, not the class. I only want the constructor method to be private which forces use of the companion object to construct the instances because the companion object is allowed to see private methods/members of the class.